gufe/storage/stagingregistry.py

```diff
@@ -0,0 +1,397 @@
+    except ValueError:
+        return True
+    else:
+        return False
+
+
+class StagingRegistry:
+    """PathLike local representation of an :class:`.ExternalStorage`.
+
+    This connects objects on a local filesystem to the key-value store of a
+    (possibly remote) :class:`.ExternalStorage`. It presents a FileLike
+    interface to users, but internally (via the :class:`.StagingPath` objects
+    it contains in its registry) maps local filenames to the keys (labels)
+    for the key-value store.
+
+    1. If a local path is requested that corresponds to an existing label in
+       the :class:`.ExternalStorage`, this object will "download" the
+       contents of that key to that local path.
+
+    2. When requested, it transfers any newly created files to the
+       :class:`.ExternalStorage`.
+
+    3. It can delete all of the files it manages.
+
+    Parameters
+    ----------
+    scratch : PathLike
+        the scratch directory shared by all objects on this host
+    external : :class:`.ExternalStorage`
+        external storage resource where objects should eventualy go
+    prefix : str
+        label for this specific unit; this should be a slash-separated
+        description of where this unit fits in the hierarchy. For example,
+        it might be ``$DAG_LABEL/$UNIT_LABEL`` or
+        ``$DAG_LABEL/$UNIT_LABEL/$UNIT_REPEAT``. It must be a unique
+        identifier for this unit within the permanent storage.
+    staging : PathLike
+        name of the subdirectory of scratch where staged results are
+        temporarily stored; default is '.staging'. This must be the same for
+        all units within a DAG.
+    delete_staging : bool
+        whether to delete the contents of the $SCRATCH/$STAGING/$PREFIX
+        directory when this object is deleted
+    """
+    def __init__(
+        self,
+        scratch: PathLike,
+        external: ExternalStorage,
+        prefix: str,
+        *,
+        staging: PathLike = Path(".staging"),
+        delete_staging: bool = True,
+        delete_empty_dirs: bool = True,
+    ):
+        self.external = external
+        self.scratch = Path(scratch)
+        self.prefix = Path(prefix)
+        self.delete_staging = delete_staging
+        self.delete_empty_dirs = delete_empty_dirs
+        self.staging = staging
+
+        self.registry: set[StagingPath] = set()
+        self.preexisting: set[StagingPath] = set()
+        self.staging_dir = self.scratch / staging / prefix
+        self.staging_dir.mkdir(exist_ok=True, parents=True)
+
+    def _delete_staging_safe(self):
+        """Check if deleting staging will remove data from external.
+        """
+        return _safe_to_delete_staging(
+            external=self.external,
+            path=self.staging_dir,
+            prefix=self.prefix,
+        )
+
+    def transfer_single_file_to_external(self, held_file: StagingPath):
+        """Transfer a given file from staging into external storage
+        """
```

Comment:
> if I've followed this, it returns the held_file back if it was transferred, but otherwise None? 

Yes. The real point here is that `transfer_staging_to_external`, which tries to transfer everything, returns the list of files that have been transferred. That was needed because it turned out that the successfully transferred files are needed by the `StorageManager` if `keep_shared is False`. I might be able to move that responsibility into a `delete_transferred` method on `StagingRegistry`, though.

> Can `store_path` ever (knowingly) fail?

I'm not sure what you mean here by "knowingly." If `store_path` does not successfully store the path, it should raise an exception. The only way it can communicate failure is via exception, and if it fails silently, that's a bug.